unity/Hello_World/Assets/Scripts/AudioCapture.cs

```diff
@@ -40,121 +44,80 @@ public class AudioCapture : MonoBehaviour
+                                            true, // Loop
+                                            recordingDuration, // Length of recording (sec)
+                                            sampleRate); // Sample rate
+        audioSource.loop = true;
 
-        Thread t = new Thread(AcceptTCPClient);
-        t.Start();
-        log.LogInfo("Waiting for audio TCP connections");
+        // Wait for recording to start
+        while ((Microphone.GetPosition(null) <= 0)) { }
+        audioSource.Play();
 
-        foreach (var device in Microphone.devices)
-        {
-            log.LogInfo("Name: " + device);
-            microphone = device;
-        }
+        // In order to capture microphone audio without hearing the playback in the headset,
+        // we need to scale the AudioSource volume down here, and then scale it back up when
+        // the audio is processed in OnAudioFilterRead.
+        // https://stackoverflow.com/questions/37787343/capture-audio-from-microphone-without-playing-it-back
+        audioSource.volume = 0.01f;
 
-        try
-        {
-            audioObject = new GameObject();
-            audioSource = audioObject.AddComponent<AudioSource>();
-            audioSource.clip = Microphone.Start(microphone, true, 1, 48000);
-            audioSource.loop = true;
-
-            while ((Microphone.GetPosition(null) <= 0)) { }
-            audioSource.Play();
-            audioSource.volume = 0.01f; // Reduce the volume so we don't hear it in the headset
-
-            AudioConfiguration ac = AudioSettings.GetConfiguration();
-            //log.LogInfo("sample rate: " + AudioSettings.outputSampleRate.ToString());
-            //log.LogInfo("speakermode: " + ac.speakerMode.ToString());
-            //log.LogInfo("dsp size: " + ac.dspBufferSize.ToString());
-        }
-        catch (Exception e)
-        {
-            log.LogInfo("Exception: " + e);
-        }
-        */
-    }
-
-    // Update is called once per frame
-    void Update()
-    {
-        if (debugString != "")
-        {
-            this.logger().LogInfo(debugString);
-        }
+        running = true;
     }
 
+    // OnAudioFilterRead is called every time an audio chunk is received (every ~20ms)
+    // from the audio clip on the audio source.
+    // Audio data is an array of floats ranging from -1 to 1.
+    // Note: This function is NOT executed on the application main thread,
+    // so use of Unity functions is not permitted.
+    // More info: https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnAudioFilterRead.html
     void OnAudioFilterRead(float[] data, int channels)
     {
+        // Wait for start function to finish
+        if (!running)
+            return;
+
         // Scale the sound up to increase the volume since we reduced it earlier by 0.01x
         float[] scaledData = new float[data.Length];
         for (int i = 0; i < data.Length; i++)
         {
             scaledData[i] = data[i] * 100;

```

Comment:
Ah, I understand now. This is the inverse of the reduced volume. Something we should be aware/careful of is value truncation/error by doing this. If we find that this is an issue, we may want to follow IEEE format modulate by a power of 2 instead of 10, so we only modify the exponent part and not the mantissa (the significant bits).